src/wwwroot/app/services/VacancyService.js

```diff
@@ -1,91 +1,182 @@
+      _UserService = UserService;
    }
 
-   getVacancies() {
-      return _HttpService.get(VACANCY_URL);
-   }
-
-   getVacancy(vacancyId) {
-      const additionalUrl = VACANCY_URL + vacancyId;
-      return _HttpService.get(additionalUrl);
-   }
-
-   _saveNewTopicsToThesaurus(data) {
-      let promises = map(data, (list, thesaurusName) => {
-         return _ThesaurusService.saveThesaurusTopics(thesaurusName, list);
+   search(condition) {
+      _LoggerService.debug('search vacancies', condition);
+      const searchUrl = 'search';
+      const additionalUrl = VACANCY_URL + searchUrl;
+      return _HttpService.post(additionalUrl, condition).then(response => {
+         return _$q.all(map(response.vacancies, this.convertFromServerFormat)).then((vacancies) => {
+            response.vacancies = vacancies;
+            return response;
+         });
       });
-      return _$q.all(promises);
    }
 
-   _onError(err) {
-      _LoggerService.error(err);
-      return _$q.reject(err);
+   convertFromServerFormat(vacancy) {
+      vacancy = _convertFromServerDates(vacancy);
+      vacancy.responsibleId = toString(vacancy.responsibleId);
+      return _$q.all([_fillThesauruses(vacancy), _fillUser(vacancy)]).then(first);
    }
 
-   saveVacancy(entity) {
-      each(DATE_TYPE, (type) => {
-         entity[type] = utils.formatDateToServer(entity[type]);
-      });
-      let mapEntity = {
-         'skills' : 'requiredSkills',
-         'tags': 'tags'
-      };
+   remove(vacancy) {
+      if (vacancy.id) {
+         const additionalUrl = VACANCY_URL + vacancy.id;
+         return _HttpService.remove(additionalUrl, vacancy);
+      } else {
+         _LoggerService.debug('Can\'t remove new vacancy', vacancy);
+         return _$q.when(true);
+      }
+   }
 
-      let data = mapValues(mapEntity, (vacancyKey) => {
-         return filter(entity[vacancyKey], {id: undefined});
-      });
+   save(vacancy) {
+      vacancy = cloneDeep(vacancy);
 
-      return this._saveNewTopicsToThesaurus(data).then((promises) => {
-         each(mapEntity, (vacancyKey, thesaurusName) => {
-            remove(entity[vacancyKey], {id: undefined});
-            entity[vacancyKey] = concat(entity[vacancyKey], promises[thesaurusName]);
-            entity[vacancyKey] = map(entity[vacancyKey], 'id');
+      return _saveNewTopics(vacancy).then((storedTopics) => {
+         each(storedTopics, (list, fieldName) => {
+            vacancy[fieldName] = union(vacancy[fieldName] || [], list);
          });
+         vacancy = _convertThesaurusToIds(vacancy);
+         vacancy = _convertToServerDates(vacancy);
 
-         if (entity.id) {
-            const additionalUrl = VACANCY_URL + entity.id;
-            return _HttpService.put(additionalUrl, entity);
+         delete vacancy.createdOn;
+         delete vacancy.responsible;
+         vacancy.languageSkill = vacancy.languageSkill || {};
+         if (isEmpty(vacancy.languageSkill) ||
```

Comment:
It is doesn't work in my situation. In one case, I get empty object, in other case, I get vacancy.languageSkill with two properties, that maybe undefined, empty string or null. It is from my experience of testing this use case.